File: instagram-handler/services/account_manager.py
import json
import os
import threading
from datetime import datetime, timezone
from typing import Any

from instagrapi import Client
from instagrapi.exceptions import (
    ChallengeRequired,
    ClientError,
    LoginRequired,
    PleaseWaitFewMinutes,
    TwoFactorRequired,
)

from services import database as db
from services.crypto import decrypt, encrypt
from services.proxy_manager import resolve_proxy

SESSIONS_DIR = os.path.join(os.path.dirname(os.path.dirname(__file__)), "data", "sessions")
os.makedirs(SESSIONS_DIR, exist_ok=True)

_clients: dict[int, Client] = {}
_lock = threading.Lock()


class AccountLoginError(Exception):
    def __init__(self, message: str, needs_2fa: bool = False):
        super().__init__(message)
        self.needs_2fa = needs_2fa


def _session_path(account_id: int) -> str:
    return os.path.join(SESSIONS_DIR, f"account_{account_id}.json")


def _build_client(proxy: str = "") -> Client:
    client = Client()
    client.delay_range = [1, 3]
    if proxy:
        client.set_proxy(proxy)
    return client


def _save_session(account_id: int, client: Client) -> None:
    settings = client.get_settings()
    db.update_account(account_id, {"session_data": encrypt(json.dumps(settings))})
    with open(_session_path(account_id), "w", encoding="utf-8") as f:
        json.dump(settings, f)


def _load_session_client(account_id: int, secrets: dict[str, Any]) -> Client | None:
    client = _build_client(resolve_proxy(secrets.get("proxy") or ""))
    settings = None
    session_json = decrypt(secrets.get("session_data") or "")
    if session_json:
        try:
            settings = json.loads(session_json)
        except json.JSONDecodeError:
            settings = None
    if not settings:
        path = _session_path(account_id)
        if os.path.exists(path):
            try:
                with open(path, encoding="utf-8") as f:
                    settings = json.load(f)
            except (json.JSONDecodeError, OSError):
                return None
    if not settings:
        return None
    try:
        client.set_settings(settings)
        client.account_info()
        return client
    except Exception:
        return None


def login_account(
    account_id: int,
    password: str | None = None,
    verification_code: str | None = None,
) -> dict[str, Any]:
    secrets = db.get_account_secrets(account_id)
    if not secrets:
        raise AccountLoginError("Account not found")

    username = secrets["username"]
    pwd = password or decrypt(secrets.get("password_enc") or "")
    if not pwd and not secrets.get("session_data"):
        raise AccountLoginError("Password required")

    client = _build_client(resolve_proxy(secrets.get("proxy") or ""))

    with _lock:
        try:
            if secrets.get("session_data") or os.path.exists(_session_path(account_id)):
                loaded = _load_session_client(account_id, secrets)
                if loaded:
                    _clients[account_id] = loaded
                    return _sync_profile(account_id, loaded)

            if verification_code:
                client.login(username, pwd, verification_code=verification_code)
            else:
                client.login(username, pwd)

            _clients[account_id] = client
            _save_session(account_id, client)
            if password:
                db.update_account(account_id, {"password_enc": encrypt(password)})
            return _sync_profile(account_id, client)

        except TwoFactorRequired:
            raise AccountLoginError("2FA code required. Enter the verification code.", needs_2fa=True)
        except ChallengeRequired:
            raise AccountLoginError("Instagram challenge required. Login manually on the app first.")
        except PleaseWaitFewMinutes as exc:
            db.update_account(account_id, {"status": "error", "last_error": str(exc)})
            raise AccountLoginError(f"Rate limited: {exc}")
        except (LoginRequired, ClientError) as exc:
            db.update_account(account_id, {"status": "error", "last_error": str(exc)})
            raise AccountLoginError(str(exc))


def _sync_profile(account_id: int, client: Client) -> dict[str, Any]:
    info = client.account_info()
    now = datetime.now(timezone.utc).isoformat()
    db.update_account(
        account_id,
        {
            "status": "active",
            "followers": info.follower_count,
            "following": info.following_count,
            "posts_count": info.media_count,
            "profile_pic": str(info.profile_pic_url) if info.profile_pic_url else "",
            "full_name": info.full_name or "",
            "is_verified": 1 if info.is_verified else 0,
            "last_login": now,
            "last_error": "",
        },
    )
    db.log_activity(account_id, "login", f"Logged in as @{info.username}")
    return db.get_account(account_id)


def get_client(account_id: int) -> Client:
    with _lock:
        if account_id in _clients:
            return _clients[account_id]
    secrets = db.get_account_secrets(account_id)
    if not secrets:
        raise AccountLoginError("Account not found")
    client = _load_session_client(account_id, secrets)
    if not client:
        raise AccountLoginError("Session expired. Please login again.")
    _clients[account_id] = client
    return client


def logout_account(account_id: int) -> None:
    with _lock:
        _clients.pop(account_id, None)
    path = _session_path(account_id)
    if os.path.exists(path):
        os.remove(path)
    db.update_account(account_id, {"status": "inactive", "session_data": ""})
    db.log_activity(account_id, "logout", "Session cleared")


def refresh_account(account_id: int) -> dict[str, Any]:
    client = get_client(account_id)
    return _sync_profile(account_id, client)


def bulk_refresh(account_ids: list[int]) -> list[dict[str, Any]]:
    results = []
    for aid in account_ids:
        try:
            account = refresh_account(aid)
            results.append({"id": aid, "success": True, "account": account})
        except Exception as exc:
            results.append({"id": aid, "success": False, "error": str(exc)})
    return results


def bulk_login(account_ids: list[int]) -> list[dict[str, Any]]:
    results = []
    for aid in account_ids:
        try:
            account = login_account(aid)
            results.append({"id": aid, "success": True, "account": account})
        except AccountLoginError as exc:
            results.append({"id": aid, "success": False, "error": str(exc), "needs_2fa": exc.needs_2fa})
        except Exception as exc:
            results.append({"id": aid, "success": False, "error": str(exc)})
    return results


def post_photo(account_id: int, image_path: str, caption: str = "") -> dict[str, Any]:
    client = get_client(account_id)
    media = client.photo_upload(image_path, caption)
    db.log_activity(account_id, "post_photo", f"Posted media {media.pk}")
    db.update_account(account_id, {"posts_count": (db.get_account(account_id) or {}).get("posts_count", 0) + 1})
    return {"media_id": str(media.pk), "code": media.code}


def import_session(account_id: int, session_settings: dict[str, Any]) -> dict[str, Any]:
    secrets = db.get_account_secrets(account_id)
    if not secrets:
        raise AccountLoginError("Account not found")
    client = _build_client(resolve_proxy(secrets.get("proxy") or ""))
    client.set_settings(session_settings)
    try:
        client.get_timeline_feed()
    except Exception as exc:
        raise AccountLoginError(f"Invalid session: {exc}")
    _clients[account_id] = client
    _save_session(account_id, client)
    return _sync_profile(account_id, client)

File: instagram-handler-php/src/AccountManager.php
<?php

declare(strict_types=1);

final class AccountManager
{
    private static function sessionPath(int $accountId): string
    {
        return SESSIONS_PATH . "/account_{$accountId}.json";
    }

    public static function login(int $accountId, ?string $password = null, ?string $verificationCode = null): array
    {
        $secrets = Database::getAccountSecrets($accountId);
        if (!$secrets) {
            throw new RuntimeException('Account not found');
        }

        $pwd = $password ?: Crypto::decrypt($secrets['password_enc'] ?? '');
        if ($pwd === '' && empty($secrets['session_data'])) {
            throw new RuntimeException('Password required');
        }

        $result = InstagramBridge::call('login', [
            'username' => $secrets['username'],
            'password' => $pwd,
            'proxy' => ProxyManager::resolveProxy($secrets['proxy'] ?? ''),
            'session_path' => self::sessionPath($accountId),
            'verification_code' => $verificationCode,
        ]);

        if (empty($result['success'])) {
            if (!empty($result['needs_2fa'])) {
                throw new RuntimeException('2FA code required. Enter the verification code.');
            }
            throw new RuntimeException($result['error'] ?? 'Login failed');
        }

        $data = $result['data'];
        if ($password) {
            Database::updateAccount($accountId, ['password_enc' => Crypto::encrypt($password)]);
        }
        if (!empty($data['session_settings'])) {
            Database::updateAccount($accountId, [
                'session_data' => Crypto::encrypt(json_encode($data['session_settings'])),
            ]);
        }

        Database::updateAccount($accountId, [
            'status' => 'active',
            'followers' => $data['followers'] ?? 0,
            'following' => $data['following'] ?? 0,
            'posts_count' => $data['posts_count'] ?? 0,
            'profile_pic' => $data['profile_pic'] ?? '',
            'full_name' => $data['full_name'] ?? '',
            'is_verified' => !empty($data['is_verified']) ? 1 : 0,
            'last_login' => Database::now(),
            'last_error' => '',
        ]);
        Database::logActivity($accountId, 'login', 'Logged in as @' . $secrets['username']);
        return Database::getAccount($accountId);
    }

    public static function logout(int $accountId): void
    {
        $path = self::sessionPath($accountId);
        if (is_file($path)) {
            unlink($path);
        }
        Database::updateAccount($accountId, ['status' => 'inactive', 'session_data' => '']);
        Database::logActivity($accountId, 'logout', 'Session cleared');
    }

    public static function refresh(int $accountId): array
    {
        return self::login($accountId);
    }

    public static function bulkLogin(array $ids): array
    {
        $results = [];
        foreach ($ids as $id) {
            try {
                $account = self::login((int) $id);
                $results[] = ['id' => (int) $id, 'success' => true, 'account' => $account];
            } catch (Throwable $e) {
                $results[] = ['id' => (int) $id, 'success' => false, 'error' => $e->getMessage()];
            }
        }
        return $results;
    }

    public static function bulkRefresh(array $ids): array
    {
        return self::bulkLogin($ids);
    }

    public static function postPhoto(int $accountId, string $imagePath, string $caption = ''): array
    {
        $secrets = Database::getAccountSecrets($accountId);
        if (!$secrets) {
            throw new RuntimeException('Account not found');
        }
        $result = InstagramBridge::call('post_photo', [
            'username' => $secrets['username'],
            'proxy' => ProxyManager::resolveProxy($secrets['proxy'] ?? ''),
            'session_path' => self::sessionPath($accountId),
            'image_path' => $imagePath,
            'caption' => $caption,
        ]);
        if (empty($result['success'])) {
            throw new RuntimeException($result['error'] ?? 'Post failed');
        }
        Database::logActivity($accountId, 'post_photo', 'Posted media');
        $account = Database::getAccount($accountId);
        Database::updateAccount($accountId, ['posts_count' => ($account['posts_count'] ?? 0) + 1]);
        return $result['data'];
    }
}
